contracts/protocol/tokenization/NTokenLiquidity.sol

```diff
@@ -0,0 +1,232 @@
+    }
+
+    /// @inheritdoc INTokenLiquidity
+    function decreaseLiquidity(
+        address user,
+        uint256 tokenId,
+        uint128 liquidityDecrease,
+        uint256 amount0Min,
+        uint256 amount1Min,
+        bool receiveEth
+    ) external onlyPool nonReentrant {
+        require(user == ownerOf(tokenId), Errors.NOT_THE_OWNER);
+
+        address positionManager = _ERC721Data.underlyingAsset;
+        if (liquidityDecrease > 0) {
+            _decreaseLiquidity(
+                positionManager,
+                tokenId,
+                liquidityDecrease,
+                amount0Min,
+                amount1Min
+            );
+        }
+
+        (address token0, address token1) = _underlyingAsset(
+            positionManager,
+            tokenId
+        );
+
+        address weth;
+        if (receiveEth) {
+            weth = _addressesProvider.getWETH();
+            receiveEth = (token0 == weth || token1 == weth);
+        }
+
+        (uint256 amount0, uint256 amount1) = _collect(
+            positionManager,
+            tokenId,
+            user,
+            receiveEth
+        );
+
+        if (receiveEth) {
+            if (amount0 > 0) {
+                transferTokenOut(user, token0, amount0, weth);
+            }
+            if (amount1 > 0) {
+                transferTokenOut(user, token1, amount1, weth);
+            }
+        }
+    }
+
+    /// @inheritdoc INTokenLiquidity
+    function increaseLiquidity(
+        address payer,
+        uint256 tokenId,
+        uint256 amountAdd0,
+        uint256 amountAdd1,
+        uint256 amount0Min,
+        uint256 amount1Min
+    ) external payable onlyPool nonReentrant {
+        address positionManager = _ERC721Data.underlyingAsset;
+
+        (address token0, address token1) = _underlyingAsset(
+            positionManager,
+            tokenId
+        );
+
+        // move underlying into this contract
+        bool token0IsETH;
+        bool token1IsETH;
+        if (msg.value > 0) {
+            address weth = _addressesProvider.getWETH();
+            token0IsETH = (token0 == weth);
+            token1IsETH = (token1 == weth);
+        }
+        if (token0IsETH) {
+            require(msg.value == amountAdd0, Errors.INVALID_AMOUNT);
+        } else {
+            IERC20(token0).safeTransferFrom(payer, address(this), amountAdd0);
+            checkAllownance(token0, positionManager);
+        }
+        if (token1IsETH) {
+            require(msg.value == amountAdd1, Errors.INVALID_AMOUNT);
+        } else {
+            IERC20(token1).safeTransferFrom(payer, address(this), amountAdd1);
+            checkAllownance(token1, positionManager);
+        }
+
+        (uint256 amount0, uint256 amount1) = _increaseLiquidity(
+            positionManager,
+            tokenId,
+            amountAdd0,
+            amountAdd1,
+            amount0Min,
+            amount1Min
+        );
+
+        // refund unused tokens
```

Comment:
if token0 & token1 are not WETH, we need to refund ETH to user when msg.value > 0.